src/axom/inlet/tests/inlet_function.cpp

```diff
@@ -321,6 +321,273 @@ TEST(inlet_function, simple_vec3_to_vec3_array_of_struct)
+
+  inlet.addFunction("foo",
+                    FunctionTag::Vector,
+                    {FunctionTag::Vector},
+                    "foo's description");
+
+  auto callable =
+    inlet["foo"].get<std::function<FunctionType::Vector(FunctionType::Vector)>>();
+
+  FunctionType::Vector input_3d({3.5, 0.5, 7.5});
+  auto result = callable(input_3d);
+
+  EXPECT_EQ(result.dim, 3);
+  EXPECT_FLOAT_EQ(result.vec[0], 7);
+  EXPECT_FLOAT_EQ(result.vec[1], 0);
+  EXPECT_FLOAT_EQ(result.vec[2], 1);
+
+  FunctionType::Vector input_2d({3.5, 0.5});
+  result = callable(input_2d);
+
+  EXPECT_EQ(result.dim, 2);
+  EXPECT_FLOAT_EQ(result.vec[0], 7);
+  EXPECT_FLOAT_EQ(result.vec[1], 1);
+}
+
+template <typename Ret, typename... Args>
+Ret checkedCall(const sol::protected_function& func, Args&&... args)
+{
+  auto tentative_result = func(std::forward<Args>(args)...);
+  EXPECT_TRUE(tentative_result.valid());
+  return tentative_result;
+}
+
+TEST(inlet_function, lua_usertype_basic)
+{
+  std::string testString = "function func(vec) return 7 end";
+  LuaReader lr;
+  lr.parseString(testString);
+  sol::protected_function func = lr.solState()["func"];
+  axom::inlet::FunctionType::Vector vec {1, 2, 3};
+  int result = checkedCall<int>(func, vec);
+  EXPECT_EQ(result, 7);
+}
+
+TEST(inlet_function, lua_usertype_basic_ret)
+{
+  std::string testString =
+    "function func(x, y, z) return Vector.new(x, y, z) end";
+  LuaReader lr;
+  lr.parseString(testString);
+  sol::protected_function func = lr.solState()["func"];
+  axom::inlet::FunctionType::Vector vec {1, 2, 3};
+  auto result = checkedCall<axom::inlet::FunctionType::Vector>(func, 1, 2, 3);
+  EXPECT_EQ(vec, result);
+}
+
+TEST(inlet_function, lua_usertype_basic_ret_2d)
+{
+  std::string testString = "function func(x, y, z) return Vector.new(x, y) end";
+  LuaReader lr;
+  lr.parseString(testString);
+  sol::protected_function func = lr.solState()["func"];
+  axom::inlet::FunctionType::Vector vec {1, 2};
+  auto result = checkedCall<axom::inlet::FunctionType::Vector>(func, 1, 2, 3);
+  EXPECT_EQ(vec, result);
+}
+
+TEST(inlet_function, lua_usertype_basic_ret_default)
+{
+  std::string testString = "function func(x, y, z) return Vector.new() end";
+  LuaReader lr;
+  lr.parseString(testString);
+  sol::protected_function func = lr.solState()["func"];
+  axom::inlet::FunctionType::Vector vec {0, 0, 0};
+  auto result = checkedCall<axom::inlet::FunctionType::Vector>(func, 1, 2, 3);
+  EXPECT_EQ(vec, result);
+}
+
+TEST(inlet_function, lua_usertype_basic_add)
+{
+  std::string testString = "function func(vec1, vec2) return vec1 + vec2 end";
+  LuaReader lr;
+  lr.parseString(testString);
+  sol::protected_function func = lr.solState()["func"];
+  axom::inlet::FunctionType::Vector vec1 {1, 2, 3};
+  axom::inlet::FunctionType::Vector vec2 {4, 5, 6};
+  auto result = checkedCall<axom::inlet::FunctionType::Vector>(func, vec1, vec2);
+  EXPECT_EQ(result, vec1.vec + vec2.vec);
```

Comment:
I like this test, but I wonder if it would be better to be explicit with the return value here, i.e.: ``Vector sum{5,7,9}``
(as opposed to relying on Vector's `operator+` in both cases).

And similarly for subtraction, negation and multiplication below.

Reply:
I went ahead and converted all the vector operation tests to use "hand-calculated" answers

Reply:
I went ahead and converted all the vector operation tests to use "hand-calculated" answers